recipes_source/recipes/tuning_guide.py

```diff
@@ -0,0 +1,376 @@
+
+for param in model.parameters():
+    param.grad = None
+
+###############################################################################
+# The second code snippet doesn't call memset on each individual parameter, also
+# the subsequent backward pass uses assignment instead of addition assignment to
+# store gradients, this reduces number of memory operations.
+
+###############################################################################
+# Fuse pointwise operations
+# ~~~~~~~~~~~~~~~~~~~~~~~~~
+# Pointwise operations (elementwise addition, multiplication, math functions -
+# ``sin()``, ``cos()``, ``sigmoid()`` etc.) should be fused into a single kernel
+# to amortize memory access time and kernel launch time.
+#
+# `PyTorch JIT <https://pytorch.org/docs/stable/jit.html>`_ can fuse kernels
+# automatically, although there could be additional fusion opportunities not yet
+# implemented in the compiler, and not all device types are supported equally.
+#
+# Unfused pointwise operations are memory-bound, for each unfused operation
+# PyTorch launches a separate kernel. Each kernel loads data from the memory,
+# performs computation (this step is usually inexpensive) and stores results
+# back into the memory.
+#
+# Fused operator launches only one kernel for multiple fused pointwise ops and
+# loads/stores data only once to the memory. This makes JIT very useful for
+# activation functions, optimizers, custom RNN cells etc.
+#
+# In the simplest case fusion can be enabled by applying
+# `torch.jit.script <https://pytorch.org/docs/stable/generated/torch.jit.script.html#torch.jit.script>`_
+# decorator to the function definition.
+# Refer to
+# `TorchScript documentation <https://pytorch.org/docs/stable/jit.html>`_
+# for more advanced use cases.
+
+###############################################################################
+# Enable channels_last memory format for computer vision models (experimental)
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# PyTorch 1.5 introduced support for `channels_last` memory format for
+# convolutional networks. This format is meant to be used in conjunction with
+# `AMP <https://pytorch.org/docs/stable/amp.html>`_ to further accelerate
+# convolutional neural networks with
+# `Tensor Cores <https://www.nvidia.com/en-us/data-center/tensor-cores/>`_.
+#
+# Support for `channels_last` is new and experimental, but it's expected to work
+# for standard computer vision models (e.g. ResNet-50, SSD). To convert models
+# to `channels_last` format follow
+# `Channels Last Memory Format Tutorial <https://pytorch.org/tutorials/intermediate/memory_format_tutorial.html>`_.
+# The tutorial includes a section on
+# `converting existing models <https://pytorch.org/tutorials/intermediate/memory_format_tutorial.html#converting-existing-models>`_.
+
+###############################################################################
+# Parallelize kernel execution
```

Comment:
Maybe we should remove this section completely for now, and reincarnate it after 1.7 with for_each operations appears? I'd hate to see people modifying their codes to do it now when in just a few weeks it will be so much better.

Reply:
I agree. If there is going to be a better way then let's remove this section for now.

Reply:
Are we planning to publicize `_foreach*` in 1.7?  I thought we were keeping it underscored for now so we can tweak the interface if we want.  (Personally, I think the leanest public interface would be: extend existing ops to accept lists of tensors/scalars, rather than exposing a new set of torch.foreach* functions)